### ISSUE TITLE
Cluster-autoscaler unnessecary scale up with pod anti-affinity

### ISSUE THREAD

**Issue by jfoechsler** [kubernetes/autoscaler#4082]
**Which component are you using?**: cluster-autoscaler

**What version of the component are you using?**: 

Component version: 1.19.1

**What k8s version are you using (`kubectl version`)?**:

<details><summary><code>kubectl version</code> Output</summary><br><pre>
$ kubectl version
Client Version: version.Info{Major:"1", Minor:"19+", GitVersion:"v1.19.6-eks-49a6c0", GitCommit:"49a6c0bf091506e7bafcdb1b142351b69363355a", GitTreeState:"clean", BuildDate:"2020-12-23T22:13:28Z", GoVersion:"go1.15.5", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"19+", GitVersion:"v1.19.8-eks-96780e", GitCommit:"96780e1b30acbf0a52c38b6030d7853e575bcdf3", GitTreeState:"clean", BuildDate:"2021-03-10T21:32:29Z", GoVersion:"go1.15.8", Compiler:"gc", Platform:"linux/amd64"}

</pre></details>

**What environment is this in?**: AWS EKS

**What did you expect to happen?**: Scale up of highest priority nodegroup only

**What happened instead?**: Scale up in higest priority followed by scale up also in other nodegroups

**How to reproduce it (as minimally and precisely as possible)**:  Scale replicas up in deployment where pods has label matching affinity rules:
```
affinity:
  podAntiAffinity:
    requiredDuringSchedulingIgnoredDuringExecution:
      - topologyKey: "kubernetes.io/hostname"
        labelSelector:
          matchExpressions:
            - key: deployment_id
              operator: In
              values:
                - "1"
```

**Anything else we need to know?**: The first scale up correctly estimates 3 nodes needed for 3 pending pods, but subsequently pods are marked unschedulable on that nodegroup resulting in further scale up in next compatible nodegroup.

```
I0515 15:02:22.870898       1 klogx.go:86] Pod ruby-user-admin-pr80/ruby-user-admin-649b5ffb77-px4ht is unschedulable
I0515 15:02:22.871065       1 klogx.go:86] Pod ruby-user-admin-pr80/ruby-user-admin-649b5ffb77-9cc2r is unschedulable
I0515 15:02:22.871504       1 klogx.go:86] Pod ruby-user-admin-pr80/ruby-user-admin-649b5ffb77-q7x6n is unschedulable
I0515 15:02:22.871795       1 scale_up.go:364] Upcoming 0 nodes
I0515 15:02:22.877449       1 priority.go:118] Successfully loaded priority configuration from configmap.
I0515 15:02:22.877489       1 priority.go:166] priority expander: eks-92bc82c0-5cd6-2b42-a82e-3c717fbb298d chosen as the highest available
I0515 15:02:22.877507       1 scale_up.go:456] Best option to resize: eks-92bc82c0-5cd6-2b42-a82e-3c717fbb298d
I0515 15:02:22.877520       1 scale_up.go:460] Estimated 3 nodes needed in eks-92bc82c0-5cd6-2b42-a82e-3c717fbb298d
I0515 15:02:22.877549       1 scale_up.go:574] Final scale-up plan: [{eks-92bc82c0-5cd6-2b42-a82e-3c717fbb298d 1->4 (max: 10)}]
I0515 15:02:22.877584       1 scale_up.go:663] Scale-up: setting group eks-92bc82c0-5cd6-2b42-a82e-3c717fbb298d size to 4
I0515 15:02:22.877616       1 auto_scaling_groups.go:219] Setting asg eks-92bc82c0-5cd6-2b42-a82e-3c717fbb298d size to 4
I0515 15:02:22.877794       1 event_sink_logging_wrapper.go:48] Event(v1.ObjectReference{Kind:"ConfigMap", Namespace:"kube-system", Name:"cluster-autoscaler-status", UID:"aa3ea0d1-bf44-4c80-9400-8ad4a57a1bbf", APIVersion:"v1", ResourceVersion:"31166445", FieldPath:""}): type: 'Normal' reason: 'ScaledUpGroup' Scale-up: setting group eks-92bc82c0-5cd6-2b42-a82e-3c717fbb298d size to 4
I0515 15:02:23.027366       1 eventing_scale_up_processor.go:47] Skipping event processing for unschedulable pods since there is a ScaleUp attempt this loop
I0515 15:02:23.028007       1 event_sink_logging_wrapper.go:48] Event(v1.ObjectReference{Kind:"Pod", Namespace:"ruby-user-admin-pr80", Name:"ruby-user-admin-649b5ffb77-px4ht", UID:"6090202e-2446-41d3-8650-319251c39746", APIVersion:"v1", ResourceVersion:"31166485", FieldPath:""}): type: 'Normal' reason: 'TriggeredScaleUp' pod triggered scale-up: [{eks-92bc82c0-5cd6-2b42-a82e-3c717fbb298d 1->4 (max: 10)}]
I0515 15:02:23.037572       1 event_sink_logging_wrapper.go:48] Event(v1.ObjectReference{Kind:"Pod", Namespace:"ruby-user-admin-pr80", Name:"ruby-user-admin-649b5ffb77-9cc2r", UID:"6e70ea57-329b-4a51-a67c-42247d638783", APIVersion:"v1", ResourceVersion:"31166474", FieldPath:""}): type: 'Normal' reason: 'TriggeredScaleUp' pod triggered scale-up: [{eks-92bc82c0-5cd6-2b42-a82e-3c717fbb298d 1->4 (max: 10)}]
I0515 15:02:23.055518       1 event_sink_logging_wrapper.go:48] Event(v1.ObjectReference{Kind:"Pod", Namespace:"ruby-user-admin-pr80", Name:"ruby-user-admin-649b5ffb77-q7x6n", UID:"7439fa76-3eb7-4543-9ba2-9831755475ba", APIVersion:"v1", ResourceVersion:"31166482", FieldPath:""}): type: 'Normal' reason: 'TriggeredScaleUp' pod triggered scale-up: [{eks-92bc82c0-5cd6-2b42-a82e-3c717fbb298d 1->4 (max: 10)}]
I0515 15:02:33.055402       1 static_autoscaler.go:229] Starting main loop
I0515 15:02:33.057271       1 filter_out_schedulable.go:65] Filtering out schedulables
I0515 15:02:33.057293       1 filter_out_schedulable.go:132] Filtered out 0 pods using hints
I0515 15:02:33.057513       1 filter_out_schedulable.go:157] Pod ruby-user-admin-pr80.ruby-user-admin-649b5ffb77-px4ht marked as unschedulable can be scheduled on node template-node-for-eks-92bc82c0-5cd6-2b42-a82e-3c717fbb298d-6643715706064775031-2. Ignoring in scale up.
I0515 15:02:33.058067       1 filter_out_schedulable.go:170] 1 pods were kept as unschedulable based on caching
I0515 15:02:33.058079       1 filter_out_schedulable.go:171] 1 pods marked as unschedulable can be scheduled.
I0515 15:02:33.058094       1 filter_out_schedulable.go:79] Schedulable pods present
I0515 15:02:33.058109       1 klogx.go:86] Pod ruby-user-admin-pr80/ruby-user-admin-649b5ffb77-9cc2r is unschedulable
I0515 15:02:33.058141       1 klogx.go:86] Pod ruby-user-admin-pr80/ruby-user-admin-649b5ffb77-q7x6n is unschedulable
I0515 15:02:33.058192       1 scale_up.go:364] Upcoming 3 nodes
I0515 15:02:33.061051       1 scale_up.go:288] Pod ruby-user-admin-649b5ffb77-9cc2r can't be scheduled on eks-92bc82c0-5cd6-2b42-a82e-3c717fbb298d, predicate checking error: node(s) didn't match pod affinity/anti-affinity, node(s) didn't match pod anti-affinity rules; predicateName=InterPodAffinity; reasons: node(s) didn't match pod affinity/anti-affinity, node(s) didn't match pod anti-affinity rules; debugInfo=
I0515 15:02:33.061100       1 scale_up.go:290] 1 other pods similar to ruby-user-admin-649b5ffb77-9cc2r can't be scheduled on eks-92bc82c0-5cd6-2b42-a82e-3c717fbb298d
I0515 15:02:33.061134       1 scale_up.go:437] No pod can fit to eks-92bc82c0-5cd6-2b42-a82e-3c717fbb298d
I0515 15:02:33.061292       1 priority.go:118] Successfully loaded priority configuration from configmap.
I0515 15:02:33.061307       1 priority.go:166] priority expander: eks-46bc82c0-5cd9-2b20-d0e9-0f190e937c02 chosen as the highest available
I0515 15:02:33.061322       1 scale_up.go:456] Best option to resize: eks-46bc82c0-5cd9-2b20-d0e9-0f190e937c02
I0515 15:02:33.061365       1 scale_up.go:460] Estimated 2 nodes needed in eks-46bc82c0-5cd9-2b20-d0e9-0f190e937c02
I0515 15:02:33.061393       1 scale_up.go:574] Final scale-up plan: [{eks-46bc82c0-5cd9-2b20-d0e9-0f190e937c02 2->4 (max: 10)}]
I0515 15:02:33.061410       1 scale_up.go:663] Scale-up: setting group eks-46bc82c0-5cd9-2b20-d0e9-0f190e937c02 size to 4
I0515 15:02:33.061465       1 auto_scaling_groups.go:219] Setting asg eks-46bc82c0-5cd9-2b20-d0e9-0f190e937c02 size to 4
I0515 15:02:33.204438       1 eventing_scale_up_processor.go:47] Skipping event processing for unschedulable pods since there is a ScaleUp attempt this loop
I0515 15:02:37.293295       1 node_instances_cache.go:156] Start refreshing cloud provider node instances cache
I0515 15:02:37.293331       1 node_instances_cache.go:168] Refresh cloud provider node instances cache finished, refresh took 9.84µs
I0515 15:02:43.216927       1 static_autoscaler.go:229] Starting main loop
I0515 15:02:43.556978       1 auto_scaling_groups.go:351] Regenerating instance to ASG map for ASGs: [eks-46bc82c0-5cd9-2b20-d0e9-0f190e937c02 eks-92bc82c0-5cd6-2b42-a82e-3c717fbb298d]
I0515 15:02:43.641808       1 auto_scaling.go:199] 2 launch configurations already in cache
I0515 15:02:43.641845       1 auto_scaling_groups.go:409] Instance group eks-92bc82c0-5cd6-2b42-a82e-3c717fbb298d has only 1 instances created while requested count is 4. Creating placeholder instance with ID i-placeholder-eks-92bc82c0-5cd6-2b42-a82e-3c717fbb298d-1.
I0515 15:02:43.641856       1 auto_scaling_groups.go:409] Instance group eks-92bc82c0-5cd6-2b42-a82e-3c717fbb298d has only 1 instances created while requested count is 4. Creating placeholder instance with ID i-placeholder-eks-92bc82c0-5cd6-2b42-a82e-3c717fbb298d-2.
I0515 15:02:43.641865       1 auto_scaling_groups.go:409] Instance group eks-92bc82c0-5cd6-2b42-a82e-3c717fbb298d has only 1 instances created while requested count is 4. Creating placeholder instance with ID i-placeholder-eks-92bc82c0-5cd6-2b42-a82e-3c717fbb298d-3.
I0515 15:02:43.643480       1 aws_manager.go:269] Refreshed ASG list, next refresh after 2021-05-15 15:03:43.643471794 +0000 UTC m=+929.297127470
I0515 15:02:43.644767       1 static_autoscaler.go:320] 5 unregistered nodes present
I0515 15:02:43.645396       1 filter_out_schedulable.go:65] Filtering out schedulables
I0515 15:02:43.645416       1 filter_out_schedulable.go:132] Filtered out 0 pods using hints
I0515 15:02:43.645549       1 filter_out_schedulable.go:157] Pod ruby-user-admin-pr80.ruby-user-admin-649b5ffb77-px4ht marked as unschedulable can be scheduled on node template-node-for-eks-46bc82c0-5cd9-2b20-d0e9-0f190e937c02-2456938384156277127-1. Ignoring in scale up.
I0515 15:02:43.645650       1 filter_out_schedulable.go:157] Pod ruby-user-admin-pr80.ruby-user-admin-649b5ffb77-9cc2r marked as unschedulable can be scheduled on node template-node-for-eks-92bc82c0-5cd6-2b42-a82e-3c717fbb298d-4724875543908344324-0. Ignoring in scale up.
I0515 15:02:43.646738       1 filter_out_schedulable.go:170] 0 pods were kept as unschedulable based on caching
I0515 15:02:43.646826       1 filter_out_schedulable.go:171] 2 pods marked as unschedulable can be scheduled.
I0515 15:02:43.648862       1 filter_out_schedulable.go:79] Schedulable pods present
I0515 15:02:43.648991       1 klogx.go:86] Pod ruby-user-admin-pr80/ruby-user-admin-649b5ffb77-q7x6n is unschedulable
I0515 15:02:43.649100       1 scale_up.go:364] Upcoming 5 nodes
I0515 15:02:43.649344       1 scale_up.go:288] Pod ruby-user-admin-649b5ffb77-q7x6n can't be scheduled on eks-46bc82c0-5cd9-2b20-d0e9-0f190e937c02, predicate checking error: node(s) didn't match pod affinity/anti-affinity, node(s) didn't match pod anti-affinity rules; predicateName=InterPodAffinity; reasons: node(s) didn't match pod affinity/anti-affinity, node(s) didn't match pod anti-affinity rules; debugInfo=
I0515 15:02:43.649431       1 scale_up.go:437] No pod can fit to eks-46bc82c0-5cd9-2b20-d0e9-0f190e937c02
I0515 15:02:43.649631       1 scale_up.go:288] Pod ruby-user-admin-649b5ffb77-q7x6n can't be scheduled on eks-92bc82c0-5cd6-2b42-a82e-3c717fbb298d, predicate checking error: node(s) didn't match pod affinity/anti-affinity, node(s) didn't match pod anti-affinity rules; predicateName=InterPodAffinity; reasons: node(s) didn't match pod affinity/anti-affinity, node(s) didn't match pod anti-affinity rules; debugInfo=
I0515 15:02:43.650180       1 scale_up.go:437] No pod can fit to eks-92bc82c0-5cd6-2b42-a82e-3c717fbb298d
```

**Comment by jclangst** [comment]
I think this is fixed by https://github.com/kubernetes/autoscaler/pull/3883 (fixed after 1.19.1) and is related to #4099.

**Comment by jfoechsler** [comment]
@jclangst Sounds promising. That commit does not appear to be in 1.20 either (and 1.20 also reproduces the same problem). Unless bugfix comes along, I will see how 1.21 works once clusters are upgraded to (hopefully) compatible Kubernetes 1.20.

**Comment by jclangst** [comment]
You are right. I mispoke. Only available in 1.21.0 and later.

I have run CA 1.21 on on k8s 1.20 and it worked alright. Slight mismatch on some of the resource versions which will cause a few non-crashing errors but it could be "functional" depending on your use case. YMMV.

Ideally I'd like to see these fixes ported to CA 1.20, but I don't know what the procedure for that is.
